src/safeposix/syscalls/fs_calls.rs

```diff
@@ -964,6 +963,454 @@ impl Cage {
+
+        //Decide how to proceed depending on the fd type.
+        //First we check in the file descriptor to handle sockets, streams, and pipes,
+        //and if it is a normal file descriptor we decrement the refcount to reflect
+        //one less reference to the file.
+        match &*filedesc_enum {
+            //if we are a socket, we dont change disk metadata
+            Stream(_) => {},
+            Socket(_) => {
+                Self::_cleanup_socket(self, &fd, false);
+                },
+            Pipe(pipe_filedesc_obj) => {
+                let pipenumber = pipe_filedesc_obj.pipe;
+                let read_references = Self::_lookup_refs_by_pipe_end(self, pipenumber, O_RDONLY);
+                let write_references = Self::_lookup_refs_by_pipe_end(self, pipenumber, O_WRONLY);
+
+                if write_references == 1 && pipe_filedesc_obj.flags == O_WRONLY {
+                    // let pipetable.pipenumber.eof = true;
+                }
+
+                if read_references + write_references == 1 {
+                    //del pipetable.pipenumber
+                }
+            },
+            //TO DO: check IS_EPOLL_FD and if true, call epoll_object_deallocator
+            File(normalfile_filedesc_obj) => {
+                let inodenum = normalfile_filedesc_obj.inode;
+                let inodeobj = mutmetadata.inodetable.get_mut(&inodenum).unwrap();
+                let mut fobjtable = FILEOBJECTTABLE.write().unwrap();
+
+                match inodeobj {
+                    Inode::File(ref mut normalfile_inode_obj) => {
+                        normalfile_inode_obj.refcount -= 1;
+
+                        //if it's not a reg file, then we have nothing to close
+                        //Inode::File is a regular file by default
+                        
+                        if normalfile_inode_obj.linkcount == 0 && normalfile_inode_obj.refcount == 0 {
+                            let sysfilename = format!("{}{}", FILEDATAPREFIX, inodenum);
+                            interface::removefile(sysfilename).unwrap();
+                            //removing the file from the entire filesystem
+                            mutmetadata.inodetable.remove(&inodenum);
+                            fobjtable.get(&inodenum).unwrap().close().unwrap();
+                            fobjtable.remove(&inodenum);  
+                        } 
+                    },
+                    Inode::Dir(ref mut dir_inode_obj) => {
+                        dir_inode_obj.refcount -= 1;
+
+                        //if it's not a reg file, then we have nothing to close
+                        match fobjtable.get(&inodenum) {
+                            Some(_) => {syscall_error(Errno::ENOEXEC, "close or dup", "Non-regular file in file object table");},
+                            None => {}
+                        }
+                        if dir_inode_obj.linkcount == 0 && dir_inode_obj.refcount == 0 {
```

Comment:
The minimum linkcount for dirs should be 2, so you should check it against 2 rather than 0

Reply:
The 2 links being .. and .